src/components/BitcoinPrice.js

```diff
@@ -0,0 +1,53 @@
+import React, { useEffect, useState } from 'react'
+
+const BitcoinPrice = () => {
+    const [bitcoinPrice, setBitcoinPrice] = useState(null);
+    const [socketError, setSocketError] = useState(null)
+
+    useEffect(() => {
+        // initialize bitcoin web socket
+        const bitcoinSocket = new WebSocket(
+          'wss://stream.binance.com:9443/ws/btcusdt@trade'
+        );
+
+        // set price on message
+        bitcoinSocket.onmessage = function (event) {
+          let messageObject = JSON.parse(event.data);
+          setBitcoinPrice(messageObject.p);
+        };
+        
+        // handle error state
+        const handleSocketError = (event) => {
+          if (event.message) {
+            setSocketError(event.message)
+          }
+        }
+        bitcoinSocket.addEventListener('error', handleSocketError);
+
+        // close socket/remove listener when this component unmounts
+        // to avoid memory leaks
+        return () => {
```

Comment:
Have you messed with these cleanup functions at the end of useEffects? 

They get called when the component unmounts. At this point in the app we are only on this one page and not routing to another page - nor is this component being unmounted... so this isn't technically useful right now.

But it's good to make sure!

Since I added a window listener for the errors, I want to remove it when this unmounts. I also want to close out the websocket so it doesn't keep firing once this component isn't used any more.

Reply:
Neat! I had not used them before. Making a note for the future!